Add Tracer operations specs. (#105)
open-telemetry/opentelemetry-specification: specification/tracing-api.md

### Tracer operations

#### `GetCurrentSpan`: returns the current Span from the current context.

There should be no parameters.

Returns the a default `Span` that does nothing and has an invalid `SpanContext` if no
`Span` is associated with the current context, otherwise the current `Span` from the context.

#### `WithSpan`: enters the scope of code where the given `Span` is in the current context.

Required parameters:

- The `Span` to be set to the current context.

Returns an object that defines a scope where the given `Span` will be set to the current context.

The scope is exited when the returned object is closed.

#### `SpanBuilder`: returns a `SpanBuilder` to create and start a new `Span`.

Required parameters:

- Name of the span.

Returns a `SpanBuilder` to create and start a new `Span`.

> #### `RecordSpanData`: records a `SpanData`.

Reviewer: Maybe we want to mention whether this function is blocking or non-blocking in the spec?

Author: Agree, added a note here.

Reviewer 2: Perhaps the same note should be on `Span.End` description.

Reviewer 2: https://github.com/open-telemetry/opentelemetry-specification/pull/107

Required parameters:

> - `SpanData` to be reported to all exporters.

Reviewer: Curious, if the purpose is to export the spandata, why not just call it "export"?

Author: My guess is the term `export` is used elsewhere such as the exporter interfaces in SDK, and we don't want to overload it.

This API allows to send a pre-populated span object to the exporter.
Sampling and recording decisions as well as other collection optimizations is a
responsibility of a caller.

Note, the `SpanContext` object on the span population with
the values that will allow correlation of telemetry is also a caller responsibility.

> #### `GetBinaryFormat`: returns the binary format of `Span`s.

Reviewer: Do we want this to be a Tracer method? It feels like a static method used by propagators.

Author: No because we want to allow different implementations for it. If no implementation is provided, API will provide a noop binary format for `GetBinaryFormat`.

Reviewer: Let's take a step back and consider the following questions:
1. What is our definition of tracer? How do we decide if something belongs to a tracer or not? (e.g. do we think propagator belongs to the tracer? if we need two propagators, what shall we do?)
2. If today we have `BinaryFormat` and `HttpTextFormat`, tomorrow we discovered that we need `MqttMetadataFormat` or `YetAnotherFormat`, do we end up adding APIs to `Tracer`?

Reviewer 3: I think this initially belongs to a `Tracer` given that you are defining propagators for `SpanContext`, which is tied to the tracing system - that being said, I'm curious about what to do if additional formats are needed (although I'm also curious about how common this is).

(Also, a related item: we had discussed the possibility of composite formats, in case you needed to use text-based formats, such as try `B2`, then fallback to `TraceContext`, then fallback to something else, etc).

Author: I think @reyang raised valid concerns. I wasn't in the previous discussions on propagators, but in OpenCensus we have an independent package [`trace.propagation`](https://github.com/census-instrumentation/opencensus-java/tree/master/api/src/main/java/io/opencensus/trace/propagation) specifically for propagators. Also we allowed getting explicit formats via API calls, like `getB3Format` vs. `getTraceContextFormat`.

Anyway I filed https://github.com/open-telemetry/opentelemetry-specification/issues/112 for further discussions.

### Tracer operations

#### `GetCurrentSpan`: returns the current Span from the current context.

There should be no parameters.

> Returns the a default `Span` that does nothing and has an invalid `SpanContext` if no

Reviewer 2: ```suggestion
Returns the default `Span` that does nothing and has an invalid `SpanContext` if no
```

> `Span` is associated with the current context, otherwise the current `Span` from the context.

Reviewer 4: If there are more than one Tracer instantiated, potentially with different Resource definitions, can there be more than one current span?  I would assume that there's a single current active span. 

Therefore, I'm not sure why this should be a Tracer method. If there was an explicit `Scope` type, as I think there should be, you'd be asking for `Scope.Active()`.

Author: > If there are more than one Tracer instantiated, potentially with different Resource definitions, can there be more than one current span? I would assume that there's a single current active span.

Agree having one active span makes more sense. Filed https://github.com/open-telemetry/opentelemetry-specification/issues/108 for discussion about where the API should live.

#### `WithSpan`: enters the scope of code where the given `Span` is in the current context.

Required parameters:

> - The `Span` to be set to the current context.

Reviewer: If user passed in a span has been closed, what would happen? Unspecified behavior?

Reviewer 3: I think so, yeah. This is responsibility of the user ;)

Reviewer 2: I think SpanContext is available from completed Span. So the only effect would be that all subsequent spans and logs are associated with the Span. All attempts to add links and attributes to the current span will not succeed. But again, it may be legitimate use case to associate telemetry with the completed span

Author: > This is responsibility of the user ;)

+1

Returns an object that defines a scope where the given `Span` will be set to the current context.

> The scope is exited when the returned object is closed.

Author: That's correct, updated.

> #### `SpanBuilder`: returns a `SpanBuilder` to create and start a new `Span`.

Reviewer 5: Looks like this (builder pattern) is not gonna applicable for all the languages (ex. NodeJS).

Author: This is being discussed at https://github.com/open-telemetry/opentelemetry-specification/issues/37. Generally we're using builder pattern to specify which parameters are required when initializing and which can be set after initialization. IMO this doesn't mean all languages need to conform to the builder pattern.

Reviewer 3: I have to say also in Python we won't be using the builder pattern - and given that enough languages won't be using this one, we should probably change this so a builder is not mentioned here at all ;)

Author: I'm good with replacing builder pattern in the specs and rephrasing it in a more language-agnostic way, as long as we make it clear which parameters are required/immutable vs. which can be mutated. I'd prefer to update it after we reach to an agreement on #37.

Required parameters:

- Name of the span.

Returns a `SpanBuilder` to create and start a new `Span`.

#### `RecordSpanData`: records a `SpanData`.

Required parameters:

- `SpanData` to be reported to all exporters.

This API allows to send a pre-populated span object to the exporter.

> Sampling and recording decisions as well as other collection optimizations is a

Reviewer 5: ```suggestion
Sampling and recording decisions as well as other collection optimizations are a responsibility of a caller.
```

responsibility of a caller.

> Note, the `SpanContext` object on the span population with

Reviewer 5: ```suggestion
Note, the `SpanContext` object in the span population with
```

the values that will allow correlation of telemetry is also a caller responsibility.

This API should be non-blocking.

#### `GetBinaryFormat`: returns the binary format of `Span`s.

> There should be no parameters.

Reviewer: no parameters -> no parameter.

Author: Fixed.

Returns the binary format for this implementation. If no implementation is provided
then no-op implementation will be used.

> #### `GetHttpTextFormat`: returns the HTTP text format of `Span`s.

Reviewer 6: Does it return an object or an interface that injects/extracts span context? OR does it return the span context in that format?

Author: It returns the format interface. Updated to make it clear.
